ingestion-sink/src/main/java/com/mozilla/telemetry/ingestion/transform/FileGroupByKey.java

```diff
@@ -0,0 +1,101 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package com.mozilla.telemetry.ingestion.transform;
+
+import com.mozilla.telemetry.ingestion.util.KV;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+public class FileGroupByKey<KeyType, OutputType>
+    implements Function<KV<KeyType, String>, CompletableFuture<OutputType>> {
+
+  private final Map<KeyType, Batch> batches = new HashMap<>();
+  private final int maxBytes;
+  private final long maxDelayMillis;
+  private final int maxMessages;
+  private final Function<KV<KeyType, File>, CompletableFuture<OutputType>> handler;
+
+  public FileGroupByKey(int maxBytes, Duration maxDelay, int maxMessages,
+      Function<KV<KeyType, File>, CompletableFuture<OutputType>> handler) {
+    this.maxBytes = maxBytes;
+    maxDelayMillis = maxDelay.toMillis();
+    this.maxMessages = maxMessages;
+    this.handler = handler;
+  }
+
+  synchronized public CompletableFuture<OutputType> apply(KV<KeyType, String> kv) {
+    return Optional.ofNullable(batches.get(kv.key)).flatMap(b -> b.add(kv.value)).orElseGet(() -> {
+      final Batch batch = new Batch(kv.key, handler);
+      final CompletableFuture<OutputType> response = batch.add(kv.value)
+          .orElseThrow(() -> new IllegalArgumentException("Empty batch rejected message"));
+      batches.put(kv.key, batch);
+      return response;
+    });
+  }
+
+  private class Batch {
+
+    private final KeyType key;
+    private final File file;
+    private final BufferedWriter writer;
+    private final CompletableFuture<Void> full;
+    private final CompletableFuture<OutputType> result;
+
+    int size = 0;
+    int byteSize = 0;
+
+    private Batch(KeyType key, Function<KV<KeyType, File>, CompletableFuture<OutputType>> handler) {
+      this.key = key;
+      try {
+        file = File.createTempFile("", "");
```

Comment:
Perhaps this is a future improvement rather than something for the initial scope of this PR, but writing the message contents to disk may end up being the performance bottleneck here.

Instead, a `Batch` instance could initiate a multi-part upload to GCS, buffer incoming messages to an in-memory writer, periodically flush the buffer to GCS, and then pass on the mulit-part upload ID, and the next step of processing could finalize the object by making a completion request to GCS with the upload ID.

At least that's an approximation of how it would work on S3. It looks like GCS doesn't have multipart upload, but instead allows you to compose objects together. It's not immediately obvious to me how you would achieve the same workflow on GCS.

Reply:
sounds good, I'll attempt that approach if disk batching proves too slow.

Reply:
I know hindsight queues everything to disk, sometimes more than once, and can generally handle all of telemetry on a single machine if it's not doing significant computation per-message. I was assuming that while an individual batch might fit in memory, all pending batches would not, but the multi-part method you specify should circumvent that.

Reply:
Good point. Definitely sounds like something to delay until we get some experience with the simpler implementation and see if a more complex solution is warranted.